File: docs/aggregator.js
// Build the daily wide frame from the streaming-aggregated parser output.
import { SUPPORTED_METRICS } from "./parser.js";

// Metrics that get a 30-day rolling baseline + z-score column. Phase 3
// Readiness/Environment scores read these baselines.
const BASELINE_METRICS = [
  "resting_hr", "hrv", "respiratory", "sleep_score", "spo2", "daylight",
  "walking_hr", "wrist_temp_delta_c",
];

// "date key" = local YYYY-MM-DD of a Date instance (used as map key)
function dateKey(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function* daysBetween(start, end) {
  const cur = new Date(start.getFullYear(), start.getMonth(), start.getDate());
  const stop = new Date(end.getFullYear(), end.getMonth(), end.getDate());
  while (cur <= stop) {
    yield dateKey(cur);
    cur.setDate(cur.getDate() + 1);
  }
}

// Reduce a streaming accumulator { sum, count, min, max, sumSq } to a single
// scalar per the spec's aggregation mode.
function finalizeAcc(acc, mode) {
  if (!acc || acc.count === 0) return undefined;
  if (mode === "sum") return acc.sum;
  if (mode === "min") return acc.min;
  if (mode === "max") return acc.max;
  return acc.sum / acc.count;  // mean (default)
}

const ASLEEP_STAGES = new Set(["Asleep", "AsleepCore", "AsleepDeep", "AsleepREM", "AsleepUnspecified"]);
const IN_BED_STAGES = new Set(["InBed"]);

function sleepDaily(sleep) {
  // Attribute each session to the date of its END timestamp (matches Apple Health UI).
  const byDate = new Map();
  for (const s of sleep) {
    const k = dateKey(s.end);
    let d = byDate.get(k);
    if (!d) {
      d = { asleepMin: 0, inBedMin: 0, deepMin: 0, remMin: 0, awakeMin: 0,
            sleepStart: null, sleepEnd: null };
      byDate.set(k, d);
    }
    const minutes = (s.end - s.start) / 60000;
    if (ASLEEP_STAGES.has(s.stage)) d.asleepMin += minutes;
    if (IN_BED_STAGES.has(s.stage)) d.inBedMin += minutes;
    if (s.stage === "AsleepDeep")   d.deepMin += minutes;
    if (s.stage === "AsleepREM")    d.remMin += minutes;
    if (s.stage === "Awake")        d.awakeMin += minutes;
    if (!d.sleepStart || s.start < d.sleepStart) d.sleepStart = s.start;
    if (!d.sleepEnd   || s.end   > d.sleepEnd)   d.sleepEnd = s.end;
  }
  return byDate;
}

// 0-100 sleep score: 50% duration vs 7.5h target + 25% deep ratio vs 20%
// + 25% REM ratio vs 25%. Mirrors the Python aggregator and parse_health.py.
function sleepScoreOf(asleepMin, deepMin, remMin) {
  const asleepH = asleepMin / 60;
  if (asleepH < 1) return null;
  const dur = Math.min(asleepH / 7.5, 1) * 50;
  const deepRatio = deepMin / asleepMin;
  const remRatio = remMin / asleepMin;
  const deep = Math.min(deepRatio / 0.20, 1) * 25;
  const rem  = Math.min(remRatio  / 0.25, 1) * 25;
  return Math.round((dur + deep + rem) * 10) / 10;
}

// Bedtime as minutes after 18:00 of the preceding evening (e.g. 23:30 → 330,
// 01:00 next morning → 420). Lets correlations treat bedtime as a scalar.
function bedtimeOffset(start) {
  if (!start) return null;
  const ref = new Date(start);
  ref.setHours(18, 0, 0, 0);
  if (start.getHours() < 18) ref.setDate(ref.getDate() - 1);
  return (start - ref) / 60000;
}

// Alternate bedtime form: 23:30 → 23.5, 01:30 → 25.5 (24+ for after-midnight).
// Matches scripts/parse_health.py canonical schema.
function bedtimeHour(start) {
  if (!start) return null;
  const h = start.getHours() + start.getMinutes() / 60;
  return Math.round((h < 12 ? h + 24 : h) * 100) / 100;
}

// SpO2 minimum among readings that fall inside any sleep session, attributed to
// the session's end-day.
function spo2DuringSleep(spo2Records, sleepRecords) {
  const sessions = sleepRecords
    .filter((s) => ASLEEP_STAGES.has(s.stage) || IN_BED_STAGES.has(s.stage))
    .sort((a, b) => a.start - b.start);
  if (!sessions.length || !spo2Records.length) return new Map();
  const out = new Map();
  let i = 0;
  for (const r of spo2Records) {
    while (i < sessions.length && sessions[i].end < r.start) i++;
    for (const j of [i - 1, i]) {
      if (j < 0 || j >= sessions.length) continue;
      const s = sessions[j];
      if (s.start <= r.start && r.start <= s.end) {
        const k = dateKey(s.end);
        const cur = out.get(k);
        out.set(k, cur === undefined ? r.value : Math.min(cur, r.value));
        break;
      }
    }
  }
  return out;
}

// Add `${col}_baseline30` and `${col}_zscore30` to the rows in-place. window=30,
// minPeriods=7 mirrors the Python aggregator so the two pipelines produce
// comparable Phase 3 inputs.
function attachBaselines(rows, columns, baselineCols, window = 30, minPeriods = 7) {
  for (const col of baselineCols) {
    if (!columns.includes(col)) continue;
    const baseKey = `${col}_baseline30`;
    const zKey = `${col}_zscore30`;
    columns.push(baseKey, zKey);
    for (let i = 0; i < rows.length; i++) {
      const lo = Math.max(0, i - window + 1);
      let n = 0, sum = 0, ss = 0;
      for (let j = lo; j <= i; j++) {
        const v = rows[j][col];
        if (Number.isFinite(v)) { n++; sum += v; ss += v * v; }
      }
      if (n < minPeriods) {
        rows[i][baseKey] = NaN;
        rows[i][zKey] = NaN;
        continue;
      }
      const mean = sum / n;
      // sample std (Bessel) to match pandas .std()
      const variance = n > 1 ? Math.max(0, (ss - n * mean * mean) / (n - 1)) : 0;
      const sd = Math.sqrt(variance);
      const v = rows[i][col];
      rows[i][baseKey] = mean;
      rows[i][zKey] = (Number.isFinite(v) && sd > 0) ? (v - mean) / sd : NaN;
    }
  }
}

export function buildDailyFrame(parsed) {
  // parsed.dailyAggs: Map<dateKey, Map<metricKey, {sum,count,min,max,sumSq}>>
  // parsed.sleep:     [{ start, end, stage, source }]
  // parsed.spo2Raw:   [{ start, value }]
  const cols = new Map(); // colKey -> Map<dateKey, value>
  const allDates = new Set();
  const dailyAggs = parsed.dailyAggs || new Map();

  // Walk the per-day accumulators and finalize each metric per its agg mode.
  // Same observable result as the old "iterate raw records, aggregate" pass,
  // but without ever holding the raw record arrays in memory.
  const presentMetrics = new Set();
  for (const [dKey, dayMap] of dailyAggs) {
    for (const mKey of dayMap.keys()) presentMetrics.add(mKey);
    allDates.add(dKey);
  }
  for (const spec of SUPPORTED_METRICS) {
    if (!presentMetrics.has(spec.key)) continue;
    const m = new Map();
    for (const [dKey, dayMap] of dailyAggs) {
      const acc = dayMap.get(spec.key);
      const v = finalizeAcc(acc, spec.agg);
      if (v !== undefined) m.set(dKey, v);
    }
    if (m.size) cols.set(spec.key, m);
  }

  // HR derivations: min / max / std / samples come straight from the HR
  // accumulator (count, min, max, sumSq are already there from streaming).
  if (presentMetrics.has("hr")) {
    const minM = new Map(), maxM = new Map(), stdM = new Map(), nM = new Map();
    for (const [dKey, dayMap] of dailyAggs) {
      const acc = dayMap.get("hr");
      if (!acc || acc.count === 0) continue;
      const mean = acc.sum / acc.count;
      const variance = acc.count > 1
        ? Math.max(0, (acc.sumSq - acc.count * mean * mean) / (acc.count - 1))
        : 0;
      minM.set(dKey, acc.min);
      maxM.set(dKey, acc.max);
      stdM.set(dKey, Math.sqrt(variance));
      nM.set(dKey, acc.count);
    }
    cols.set("heart_rate_min", minM);
    cols.set("heart_rate_max", maxM);
    cols.set("heart_rate_std", stdM);
    cols.set("heart_rate_samples", nM);
  }

  if (parsed.sleep && parsed.sleep.length) {
    const s = sleepDaily(parsed.sleep);
    const sleepHours = new Map();
    const deep = new Map(), rem = new Map(), awake = new Map();
    const eff = new Map(), bed = new Map(), bedHr = new Map(), score = new Map();
    for (const [k, d] of s) {
      const minutes = d.asleepMin || d.inBedMin;
      sleepHours.set(k, minutes / 60);
      deep.set(k, d.deepMin);
      rem.set(k, d.remMin);
      awake.set(k, d.awakeMin);
      if (d.inBedMin > 0) eff.set(k, d.asleepMin / d.inBedMin);
      const off = bedtimeOffset(d.sleepStart);
      if (off != null) bed.set(k, off);
      const bh = bedtimeHour(d.sleepStart);
      if (bh != null) bedHr.set(k, bh);
      const sc = sleepScoreOf(d.asleepMin, d.deepMin, d.remMin);
      if (sc != null) score.set(k, sc);
      allDates.add(k);
    }
    cols.set("sleep_hours", sleepHours);
    cols.set("sleep_deep_minutes", deep);
    cols.set("sleep_rem_minutes", rem);
    cols.set("sleep_awake_minutes", awake);
    cols.set("sleep_efficiency", eff);
    cols.set("bedtime_offset_min", bed);
    cols.set("bedtime_hour", bedHr);
    cols.set("sleep_score", score);

    // SpO2 sleep min: uses spo2Raw (kept full from streaming, small set) and
    // the sleep windows. Interval join — see spo2DuringSleep below.
    if (parsed.spo2Raw && parsed.spo2Raw.length) {
      const sm = spo2DuringSleep(parsed.spo2Raw, parsed.sleep);
      if (sm.size) {
        cols.set("spo2_sleep_min", sm);
        for (const k of sm.keys()) allDates.add(k);
      }
    }
  }

  // Build dense daily frame: one row per calendar day in [min, max], NaN-fill.
  const dates = [...allDates].sort();
  if (!dates.length) return { dates: [], rows: [], columns: [] };
  const minDate = new Date(dates[0]);
  const maxDate = new Date(dates[dates.length - 1]);
  const denseDates = [...daysBetween(minDate, maxDate)];

  const columns = [...cols.keys()];
  const rows = denseDates.map((dk) => {
    const row = { date: dk };
    for (const c of columns) {
      const v = cols.get(c).get(dk);
      row[c] = v === undefined ? NaN : v;
    }
    return row;
  });

  attachBaselines(rows, columns, BASELINE_METRICS);

  return { dates: denseDates, rows, columns };
}

// Convenience: extract a column as an array aligned with rows
export function columnValues(frame, col) {
  return frame.rows.map((r) => r[col]);
}

// Rolling mean ignoring NaN; window is centered-trailing (right-aligned).
export function rollingMean(values, window) {
  const out = new Array(values.length).fill(NaN);
  const minPeriods = Math.max(2, Math.floor(window / 2));
  for (let i = 0; i < values.length; i++) {
    const lo = Math.max(0, i - window + 1);
    let sum = 0, n = 0;
    for (let j = lo; j <= i; j++) {
      const v = values[j];
      if (Number.isFinite(v)) { sum += v; n++; }
    }
    out[i] = n >= minPeriods ? sum / n : NaN;
  }
  return out;
}

File: healthkit/aggregator.py
"""Roll the per-record DataFrames into a single daily frame for analysis."""
from __future__ import annotations

import numpy as np
import pandas as pd

from .parser import SUPPORTED_TYPES, metric_index


# Metrics that get a 30-day rolling baseline + z-score column. Picked because
# Phase 3 Readiness/Environment scores read these baselines; adding more is cheap.
BASELINE_METRICS = (
    "resting_hr", "hrv", "respiratory", "sleep_score", "spo2", "daylight",
    "walking_hr", "wrist_temp_delta_c",
)


# Sleep "day" rule: a session that ends after noon belongs to that calendar day;
# one ending at e.g. 06:00 belongs to the same calendar day. We attribute each
# session to the date of its END timestamp, which matches how Apple Health UI
# groups sleep.
def _sleep_daily(sleep: pd.DataFrame) -> pd.DataFrame:
    if sleep.empty:
        return pd.DataFrame()

    s = sleep.copy()
    s["date"] = s["end"].dt.normalize()
    s["minutes"] = (s["end"] - s["start"]).dt.total_seconds() / 60.0

    asleep_stages = {"Asleep", "AsleepCore", "AsleepDeep", "AsleepREM", "AsleepUnspecified"}
    in_bed_stages = {"InBed"}

    s["is_asleep"] = s["stage"].isin(asleep_stages)
    s["is_in_bed"] = s["stage"].isin(in_bed_stages)
    s["is_deep"] = s["stage"] == "AsleepDeep"
    s["is_rem"] = s["stage"] == "AsleepREM"
    s["is_awake"] = s["stage"] == "Awake"

    grouped = s.groupby("date")
    daily = pd.DataFrame({
        "sleep_minutes": grouped.apply(lambda g: g.loc[g["is_asleep"], "minutes"].sum()
                                       or g.loc[g["is_in_bed"], "minutes"].sum()),
        "sleep_deep_minutes": grouped.apply(lambda g: g.loc[g["is_deep"], "minutes"].sum()),
        "sleep_rem_minutes": grouped.apply(lambda g: g.loc[g["is_rem"], "minutes"].sum()),
        "sleep_awake_minutes": grouped.apply(lambda g: g.loc[g["is_awake"], "minutes"].sum()),
        "sleep_start": grouped.apply(lambda g: g["start"].min()),
        "sleep_end": grouped.apply(lambda g: g["end"].max()),
    })
    daily["sleep_hours"] = daily["sleep_minutes"] / 60.0

    # 0-100 sleep score: 50% duration vs 7.5h target + 25% deep ratio vs 20%
    # + 25% REM ratio vs 25%. Mirrors scripts/parse_health.py so the standalone
    # CLI and the library produce comparable values.
    deep_h = daily["sleep_deep_minutes"] / 60.0
    rem_h = daily["sleep_rem_minutes"] / 60.0
    asleep_h = daily["sleep_hours"]
    with np.errstate(divide="ignore", invalid="ignore"):
        dur_score = np.minimum(asleep_h / 7.5, 1.0) * 50
        deep_ratio = np.where(asleep_h > 0, deep_h / asleep_h, 0)
        rem_ratio = np.where(asleep_h > 0, rem_h / asleep_h, 0)
        deep_score = np.minimum(deep_ratio / 0.20, 1.0) * 25
        rem_score = np.minimum(rem_ratio / 0.25, 1.0) * 25
    score = pd.Series(dur_score + deep_score + rem_score, index=daily.index)
    daily["sleep_score"] = score.where(asleep_h >= 1.0).round(1)

    # bedtime as minutes-after-18:00 so 23:30 -> 330, 01:00 -> 420
    def _bedtime_offset(ts: pd.Timestamp) -> float:
        if pd.isna(ts):
            return np.nan
        ref = ts.normalize() if ts.hour >= 18 else ts.normalize() - pd.Timedelta(days=1)
        ref = ref + pd.Timedelta(hours=18)
        return (ts - ref).total_seconds() / 60.0

    daily["bedtime_offset_min"] = daily["sleep_start"].map(_bedtime_offset)

    # bedtime_hour: 23:30 → 23.5、01:30 → 25.5（24+ 形式方便當作 scalar 用）
    def _bedtime_hour(ts: pd.Timestamp) -> float:
        if pd.isna(ts):
            return np.nan
        h = ts.hour + ts.minute / 60.0
        return h + 24 if h < 12 else h
    daily["bedtime_hour"] = daily["sleep_start"].map(_bedtime_hour).round(2)

    # sleep efficiency if we have both in-bed and asleep records
    in_bed_min = grouped.apply(lambda g: g.loc[g["is_in_bed"], "minutes"].sum())
    daily["sleep_efficiency"] = np.where(
        in_bed_min > 0, daily["sleep_minutes"] / in_bed_min, np.nan
    )
    return daily


def _quantity_daily(key: str, df: pd.DataFrame) -> pd.Series:
    spec = metric_index()[key]
    s = df.copy()
    s["date"] = s["start"].dt.normalize()
    grouped = s.groupby("date")["value"]
    if spec.aggregation == "sum":
        return grouped.sum()
    if spec.aggregation == "mean":
        return grouped.mean()
    if spec.aggregation == "min":
        return grouped.min()
    if spec.aggregation == "max":
        return grouped.max()
    return grouped.mean()


def _hr_derived_daily(hr_df: pd.DataFrame) -> pd.DataFrame:
    """Per-day min / max / std / sample-count for raw heart-rate records.

    Mean is already computed via the standard 'hr' rollup; these companions let
    downstream analyses look at HR spread (training intensity proxy) without
    keeping every raw sample around.
    """
    s = hr_df.copy()
    s["date"] = s["start"].dt.normalize()
    grouped = s.groupby("date")["value"]
    out = pd.DataFrame({
        "heart_rate_min": grouped.min(),
        "heart_rate_max": grouped.max(),
        "heart_rate_std": grouped.std(ddof=1),
        "heart_rate_samples": grouped.count().astype("int64"),
    })
    return out


def _attach_baselines(daily: pd.DataFrame, cols: tuple[str, ...] = BASELINE_METRICS,
                     window: int = 30, min_periods: int = 7) -> None:
    """Add `{col}_baseline30` and `{col}_zscore30` for each metric in cols."""
    for col in cols:
        if col not in daily.columns:
            continue
        s = daily[col]
        mean = s.rolling(window, min_periods=min_periods).mean()
        std = s.rolling(window, min_periods=min_periods).std(ddof=1)
        daily[f"{col}_baseline30"] = mean
        daily[f"{col}_zscore30"] = (s - mean) / std.replace(0, np.nan)


def build_daily_frame(per_metric: dict[str, pd.DataFrame]) -> pd.DataFrame:
    """Combine all metrics into one DataFrame indexed by date."""
    columns: dict[str, pd.Series] = {}

    for spec in SUPPORTED_TYPES:
        df = per_metric.get(spec.key)
        if df is None or df.empty:
            continue
        columns[spec.key] = _quantity_daily(spec.key, df)

    sleep = per_metric.get("sleep")
    if sleep is not None and not sleep.empty:
        sleep_daily = _sleep_daily(sleep)
        for col in sleep_daily.columns:
            columns[col] = sleep_daily[col]

    if not columns:
        return pd.DataFrame()

    daily = pd.concat(columns, axis=1)
    daily.index = pd.to_datetime(daily.index).normalize()
    daily.index.name = "date"
    daily = daily.sort_index()

    # HR derivations (min/max/std/samples) from raw records
    hr_df = per_metric.get("hr")
    if hr_df is not None and not hr_df.empty:
        hr_extra = _hr_derived_daily(hr_df)
        hr_extra.index = pd.to_datetime(hr_extra.index).normalize()
        for col in hr_extra.columns:
            daily[col] = hr_extra[col]

    # derived: nighttime SpO2 minimum approximated via daily min if records overlap sleep
    spo2_df = per_metric.get("spo2")
    sleep_df = per_metric.get("sleep")
    if spo2_df is not None and sleep_df is not None and not spo2_df.empty and not sleep_df.empty:
        daily["spo2_sleep_min"] = _spo2_during_sleep(spo2_df, sleep_df)

    _attach_baselines(daily)

    return daily


def _spo2_during_sleep(spo2: pd.DataFrame, sleep: pd.DataFrame) -> pd.Series:
    """For each sleep day, find min SpO2 reading taken during a sleep window."""
    asleep_stages = {"Asleep", "AsleepCore", "AsleepDeep", "AsleepREM", "AsleepUnspecified", "InBed"}
    sessions = sleep[sleep["stage"].isin(asleep_stages)].copy()
    if sessions.empty:
        return pd.Series(dtype=float)

    sessions["date"] = sessions["end"].dt.normalize()
    # interval join: for each spo2 reading, find sleep session covering it
    spo2_sorted = spo2.sort_values("start").reset_index(drop=True)
    sessions_sorted = sessions.sort_values("start").reset_index(drop=True)

    # naive but bounded: use searchsorted on session starts
    starts = sessions_sorted["start"].values
    ends = sessions_sorted["end"].values
    dates = sessions_sorted["date"].values

    out: dict[pd.Timestamp, float] = {}
    idx = 0
    for ts, val in zip(spo2_sorted["start"].values, spo2_sorted["value"].values):
        while idx < len(starts) and ends[idx] < ts:
            idx += 1
        # try this and the previous session
        for j in (idx - 1, idx):
            if 0 <= j < len(starts) and starts[j] <= ts <= ends[j]:
                d = pd.Timestamp(dates[j])
                cur = out.get(d)
                out[d] = val if cur is None else min(cur, val)
                break
    if not out:
        return pd.Series(dtype=float)
    return pd.Series(out).sort_index()
